server/src/main/java/io/crate/execution/engine/aggregation/impl/TopKAggregation.java

```diff
@@ -0,0 +1,282 @@
+
+            } else if (args.length == 1) {
+                // We don't have a limit provided, let's go with the default values
+                ramAccounting.addBytes(calculateRamUsage(DEFAULT_MAP_SIZE));
+                state = new TopKState(new ItemsSketch<>(DEFAULT_MAP_SIZE), DEFAULT_LIMIT);
+            }
+            ramAccounting.addBytes(RamUsageEstimator.shallowSizeOf(state));
+        }
+        if (state instanceof TopKState topKState) {
+            topKState.itemsSketch.update(value);
+        }
+        return state;
+    }
+
+    private static int maxMapSize(int x) {
+        // max map size should be 4 * the limit based on the power of 2 to avoid errors
+        return (int) Math.pow(2, Math.ceil(Math.log(x) / Math.log(2))) * 4;
+    }
+
+    private static long calculateRamUsage(long maxMapSize) {
+        // The internal memory space usage of item sketch will never exceed 18 * maxMapSize bytes, plus a small
+        // constant number of additional bytes.
+        // https://datasketches.apache.org/docs/Frequency/FrequentItemsOverview.html
+        return maxMapSize * 18L;
+    }
+
+    @Override
+    public State reduce(RamAccounting ramAccounting, State state1, State state2) {
+        if (state1 instanceof TopKState t1 && state2 instanceof TopKState t2) {
+            return new TopKState(t1.itemsSketch.merge(t2.itemsSketch), t1.limit);
+        } else if (state1 instanceof Empty) {
+            return state2;
+        } else {
+            return state1;
+        }
+    }
+
+    @Override
+    public List<Object> terminatePartial(RamAccounting ramAccounting, State state) {
+        if (state instanceof TopKState topKState) {
+            ItemsSketch.Row<Object>[] frequentItems = topKState.itemsSketch.getFrequentItems(ErrorType.NO_FALSE_NEGATIVES);
+            int limit = Math.min(frequentItems.length, topKState.limit);
+            var result = new ArrayList<>(limit);
+            for (int i = 0; i < limit; i++) {
+                var item = frequentItems[i];
+                result.add(Map.of("item", item.getItem(), "frequency", item.getEstimate()));
+            }
+            return result;
+        }
+        return List.of();
+    }
+
+    public DataType<?> partialType() {
+        return new StateType<>(boundSignature.argTypes().getFirst());
+    }
+
+    sealed interface State {
+        State EMPTY = new Empty();
+    }
+
+    record Empty() implements State { }
+
+    record TopKState(ItemsSketch<Object> itemsSketch, int limit) implements State { }
+
+    static final class StateType<T> extends DataType<State> implements Streamer<State> {
+
+        public static final int ID = 4232;
+        private final DataType<?> innerType;
+
+        public StateType(DataType<?> innerType) {
+            this.innerType = innerType;
+        }
+
+        public StateType(StreamInput streamInput) throws IOException {
+            this.innerType = DataTypes.fromStream(streamInput);
+        }
+
```

Comment:
It also requires a `writeTo` override to match this.

Looks like our aggregation test framework actually doesn't exercise this path?